File: WebDriverTests_Java/WebDriverTests/src/test/java/core/systemControls/Select.java
package core.systemControls;


import org.openqa.selenium.By;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.Quotes;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.StringTokenizer;

public class Select extends HtmlControl
{
    private StaticText selectedText;
    private StaticText selectedText02;
    private boolean isMulti;

    public Select(By locator)
    {
        this(locator, null);
    }

    public Select(By locator, HtmlControl container)
    {
        super(locator, container);
        HtmlControlContainer controlContainer = new HtmlControlContainer(locator, container);

        selectedText = new StaticText(By.xpath("option[@selected]"), controlContainer);
        selectedText02 = new StaticText(By.xpath("option"), controlContainer);

    }



    @Override
    public String getText()
    {
        return selectedText02.getText();
    }

    @Override
    public void WaitText(String expectedText)
    {
        String description = formatWithLocator(String.format("Ожидание появления текста '%1$s' в элементе", expectedText));
        Waiter.Wait(() -> selectedText02.getText().equals(expectedText), description);

    }

    public boolean isMultiple() {
        return this.isMulti;
    }


    public List<WebElement> getOptions() {
        return this.element.findElements(By.tagName("option"));
    }

    public List<WebElement> getAllSelectedOptions() {
        ArrayList toReturn = new ArrayList();
        Iterator var2 = this.getOptions().iterator();

        while(var2.hasNext()) {
            WebElement option = (WebElement)var2.next();
            if(option.isSelected()) {
                toReturn.add(option);
            }
        }

        return toReturn;
    }


    public void selectByValue(String value) {
        List options = this.element.findElements(By.xpath(".//option[@value = " + Quotes.escape(value) + "]"));
        boolean matched = false;

        for(Iterator var4 = options.iterator(); var4.hasNext(); matched = true) {
            WebElement option = (WebElement)var4.next();
            this.setSelected(option);
            if(!this.isMultiple()) {
                return;
            }
        }

        if(!matched) {
            throw new NoSuchElementException("Cannot locate option with value: " + value);
        }
    }
    public void selectByText(String text)
    {
        Waiter.WaitExeption(() -> element.click(), "Ждем кликабельности элемента");
//        List options = this.element.findElements(By.xpath(".//option[normalize-space(.) = " + Quotes.escape(text) + "]"));
        WebElement options = this.element.findElement(By.xpath(".//option[normalize-space(.) = " + Quotes.escape(text) + "]"));
        options.click();
//        String value = (String) javascriptExecute("return $('[data-unique=\"modal-shop-warehouse\"]').val();");
        String value = (String) javascriptExecuteWebElement("return arguments[0].value;");
        WebElement optionsValue = this.element.findElement(By.xpath(".//option[@value = " + Quotes.escape(value) + "]"));
        if (!optionsValue.getText().equals(text))
            throw new NoSuchElementException("Cannot locate element with text: " + text);
    }
    public void selectByTextContains(String text)
    {
        Waiter.WaitExeption(() -> element.click(), "Ждем кликабельности элемента");

        List options = this.element.findElements(By.xpath(".//option"));
        for (int i= 0; i < options.size(); i++ )
        {
            WebElement option = (WebElement) options.get(i);
            if (option.getText().contains(text))
            {
                option.click();
                String value = (String) javascriptExecuteWebElement("return arguments[0].value;");
                WebElement optionsValue = this.element.findElement(By.xpath(".//option[@value = " + Quotes.escape(value) + "]"));
                if (!optionsValue.getText().contains(text))
                    throw new NoSuchElementException("Cannot locate element with text: " + text);
                break;
            }
            throw new NoSuchElementException("Cannot locate element with text: " + text);
        }

    }
    private void setSelected(WebElement option) {
        if(!option.isSelected()) {
            option.click();
        }}


    private String getLongestSubstringWithoutSpace(String s) {
        String result = "";
        StringTokenizer st = new StringTokenizer(s, " ");

        while(st.hasMoreTokens()) {
            String t = st.nextToken();
            if(t.length() > result.length()) {
                result = t;
            }
        }

        return result;
    }


}

File: WebDriverTests_Csharp/Autotest.Core/Autotests.Utilities/WebTestCore/SystemControls/Select.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using Autotests.Utilities.WebTestCore.TestSystem;
using OpenQA.Selenium;

namespace Autotests.Utilities.WebTestCore.SystemControls
{
    public class Select : HtmlControl
    {
        private readonly StaticText selectedText;
        private readonly StaticText selectedText02;

        public Select(By locator, HtmlControl container = null)
            : base(locator, container)
        {
            var controlContainer = new HtmlControlContainer(locator, container);
            
            selectedText = new StaticText(By.XPath("option[@selected]"), controlContainer);
            selectedText02 = new StaticText(By.XPath("option"), controlContainer);
        }
        public bool IsMultiple { get; private set; }

        public override string GetText()
        {
            return selectedText02.GetText();
        }

        public override void WaitText(string expectedText)
        {
            string description =
                FormatWithLocator(string.Format("Ожидание появления текста '{0}' в элементе", expectedText));
            Waiter.Wait(() => selectedText.GetText() == expectedText, description);
        }


        public void SelectValue(string value)
        {
            WaitEnabled();
            if (GetText() == value)
                return;
            IList<IWebElement> list = element.FindElements(By.XPath(".//option"));
            foreach (IWebElement option in list)
            {
                if (option.Text == value)
                {
                    SetSelected(option);
                    Click();
                }
            }
        }


        public void SelectByText(string text)
        {
            if (text == null)
                throw new ArgumentNullException("text", "text must not be null");
//            IList<IWebElement> list = element.FindElements(By.XPath(".//option"));
            IList<IWebElement> list = element.FindElements(By.XPath(".//option[normalize-space(.) = " + EscapeQuotes(text) + "]"));
            bool flag = false;
            foreach (IWebElement option in list)
            {
                SetSelected(option);
                if (!IsMultiple)
                    return;
                flag = true;
            }
            if (list.Count == 0 && text.Contains(" "))
            {
                string substringWithoutSpace = GetLongestSubstringWithoutSpace(text);
                foreach (
                    IWebElement option in
                        !string.IsNullOrEmpty(substringWithoutSpace)
                            ? (IEnumerable<IWebElement>)
                                element.FindElements(
                                    By.XPath(".//option[contains(., " +
                                            EscapeQuotes(substringWithoutSpace) + ")]"))
                            : (IEnumerable<IWebElement>) element.FindElements(By.TagName("option")))
                {
                    if (text == option.Text)
                    {
                        SetSelected(option);
                        if (!IsMultiple)
                            return;
                        flag = true;
                    }
                }
            }
            if (!flag)
                throw new NoSuchElementException("Cannot locate element with text: " + text);
        }

        private static string EscapeQuotes(string toEscape)
        {
            if (toEscape.IndexOf("\"", StringComparison.OrdinalIgnoreCase) > -1 && toEscape.IndexOf("'", StringComparison.OrdinalIgnoreCase) > -1)
            {
                bool flag = false;
                if (toEscape.IndexOf("\"", StringComparison.OrdinalIgnoreCase) == toEscape.Length - 1)
                    flag = true;
                string[] strArray = toEscape.Split(new char[1]
        {
          '"'
        });
                StringBuilder stringBuilder = new StringBuilder("concat(");
                for (int index = 0; index < strArray.Length; ++index)
                {
                    stringBuilder.Append("\"").Append(strArray[index]).Append("\"");
                    if (index == strArray.Length - 1)
                    {
                        if (flag)
                            stringBuilder.Append(", '\"')");
                        else
                            stringBuilder.Append(")");
                    }
                    else
                        stringBuilder.Append(", '\"', ");
                }
                return ((object)stringBuilder).ToString();
            }
            else if (toEscape.IndexOf("\"", StringComparison.OrdinalIgnoreCase) > -1)
                return string.Format(CultureInfo.InvariantCulture, "'{0}'", new object[1]
        {
           toEscape
        });
            else
                return string.Format(CultureInfo.InvariantCulture, "\"{0}\"", new object[1]
        {
          toEscape
        });
        }

        private static string GetLongestSubstringWithoutSpace(string s)
        {
            string str1 = string.Empty;
            string str2 = s;
            var chArray = new char[1]
            {
                ' '
            };
            foreach (string str3 in str2.Split(chArray))
            {
                if (str3.Length > str1.Length)
                    str1 = str3;
            }
            return str1;
        }

        private static void SetSelected(IWebElement option)
        {
            if (option.Selected)
                return;
            option.Click();
        }
    }
}